### ISSUE TITLE
Docker for AWS doesn't appear to benefit from ECR Policy during Stack / Service Creation

### ISSUE THREAD

**Issue by ambrons** [docker/for-aws#5]
### Expected behavior

When adding Amazon Policy "AmazonEC2ContainerRegistryReadOnly" to the created cloud formation from Docker for AWS this should allow docker find images without having to use `aws ecr get-login`.  Future updates of the service via the scheduler should also be able to access ECR without needing cached credentials via the `--with-registry-auth` command line argument.

This would give docker swarm parity with ECS provided by Amazon when using ECR as a private registry among not being confusing that Docker for AWS uses _some_ policies, but ignores others.

### Actual behavior

Using `docker stack deploy -c <composefile.yaml>  <stack name>` with added Amazon Policy "AmazonEC2ContainerRegistryReadOnly" results in the image not being able to be found:

```
$ docker-swarm service ps 3sgzt87ip19t
ID            NAME                   IMAGE                                                                            NODE                           DESIRED STATE  CURRENT STATE            ERROR                             PORTS
x6l8jtn9hnw0  esports_wallets.1      ###.dkr.ecr.us-east-1.amazonaws.com/path/to/image:tag  ip-172-31-20-65.ec2.internal   Ready          Rejected 2 seconds ago   "No such image: ###.d…" 
```

Using `aws ecr get-login` works during the creation using the following command form `docker stack deploy -c <composefile.yaml> --with-registry-auth <stack name>`, however after the Amazon password expires (12 hours) any mangers or nodes not primed will not be able to find the image as shown above.

### Information

Using Docker for AWS just release with docker version 1.13.0. I'm able to use `docker stack deploy -c <composefile.yaml> --with-registry-auth <stack name>` without issue having done a aws ecr get-login... prior.

However because AWS ECR docker login passwords are only good for 12 hours if the scheduler rebalances the deployed services after 12 hours and the node it's being deployed on doesn't have the existing authentication it fails to authenticate and you fail to find the image.

I've read through the issues and found this issue talking about enhancements to how --with-registry-auth works to allow it to use refresh tokens or the like. https://github.com/docker/docker/issues/24940

I choose to create a new ticket as https://github.com/docker/docker/issues/24940 appears to be more geared towards solving dockerhub and less about ECR.  I also opened this at https://github.com/docker/docker/issues/30713, but it was requested that I reopen the issue here.

I've also looked at several posts on the docker for aws forums. The one that pin points the problem with no real solution is this:

https://forums.docker.com/t/possible-to-use-aws-ecr-for-image-registry/22295/3

Considering this is a Docker for AWS and you have developed Docker AWS containers to run on the EC2 instances specific to AWS, why can't we just include the ECR ReadOnly Policy to the ProxyRole created during cloud formation to have access to ECR without having to worry about authentication at all?

I have attempted to add the Policy (AmazonEC2ContainerRegistryReadOnly) to the created ProxyRole used by the InstanceProfile for all Docker Nodes in the cloud formation, however this has zero effect. The result from a service ps is as follows:

```
$ docker-swarm service ps 3sgzt87ip19t
ID            NAME                   IMAGE                                                                            NODE                           DESIRED STATE  CURRENT STATE            ERROR                             PORTS
x6l8jtn9hnw0  esports_wallets.1      ###.dkr.ecr.us-east-1.amazonaws.com/path/to/image:tag  ip-172-31-20-65.ec2.internal   Ready          Rejected 2 seconds ago   "No such image: ###.d…"
```

However if I remove the service and create it after typing $(aws ecr get-login --region us-east-1) it works fine, finds the image and runs. However then it's back to 12 hours later I'm likely in the state referenced in the forum link above or in issue #24940.

We've been using ECS for over a year and the behavior of making sure the Instance profile role has the AmazonEC2ContainerRegistryReadOnly policy is all that's needed.

I dug a little down in the dockers deployed on an EC2 instance, cool stuff there how the SSH is actually a docker as well as the metadata service. My assumption is we're not trying to assume the instance policy role during a registry fetch. Or I could be talking out my butt there.

I've even gone as far as modifying the cloud formation template to include the ECR policies at creation time, but made no difference.

### Steps to reproduce the behavior

1. Use AWS Console to apply AmazonEC2ContainerRegistryReadOnly policy to cloud formation generated ProxyRole used by InstancePolicy of Docker For AWS
2. Create service or stack with an image referenced in ECR
3. service ps or container inspect to see that image is not found aka authentication wasn't present.

### Output of docker version:

Docker version 1.13.0, build 49bf474

### Output of docker info:

Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 6
Server Version: 1.13.0
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: awslogs
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: active
 NodeID: 6kup6o2m8rmgrydrj9k3p1089
 Is Manager: true
 ClusterID: sgq7930ifiv21xlc94jci3tsd
 Managers: 1
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.31.40.242
 Manager Addresses:
  172.31.40.242:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.4-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.785 GiB
Name: ip-172-31-40-242.ec2.internal
ID: REJ6:Q2GX:RYQD:KC3A:2ZK2:5JZJ:PT42:BT6B:64TN:XQED:WKBH:H7PC
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 77
 Goroutines: 190
 System Time: 2017-02-03T14:48:08.410775982Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

### Additional environment details (AWS, VirtualBox, physical, etc.):

AWS / Docker for AWS 1 manager and 3 workers

**Comment by kencochrane** [comment]
@ambrons so to sum up the issue it looks like there are 2 problems.
1. Need the IAM permissions to get the ecr tokens (http://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr_managed_policies.html#AmazonEC2ContainerRegistryReadOnly)
2. The ecr tokens expire every 12 hours, so we need a way to update these tokens before they expire.

Did I miss anything?

**Comment by ambrons** [comment]
@kencochrane 

Sorry for my long winded report I could see how it could be a little confusing.

The primary concern is your number one. I'm not familiar with the internals of how you are handling EC2 Instance Profiles => AWS Roles, so perhaps it is to _get the ecr tokens_ however in my AWS experience the EC2 instance InstanceProfile would be assigned a Role which attached policies.  And the EC2 instance can assume the role for the purpose of obtaining access to AWS resources. 

So in summary the ideal would be allowing the user of Docker for AWS or default inclusion of ECR Policy in CloudFormation would give Docker Managers and Nodes access to communicate with ECR without the need of the docker user having to use docker-login at all.  It would behave from the docker user's perspective as if it was a public repo.  I now ECS does this seemlessly, but I'm not sure if there's some magic they do under the covers to assume the role on the EC2 instance during a docker pull.

In either case if the above is done, the tokens expiring is moot because all instances in the Docker for AWS would have access to ECR all the time based on the above Role / Polices.

**Comment by kencochrane** [comment]
@ambrons thank you for the details. This is what I'm thinking, let me know if that works for you.

We add a new template question that asks if you want to enable ECR support (yes or no, default to no)

if they say 'no':
- do nothing, same as today.

If they say 'yes':

1.  we add the Read only profile detailed here (http://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr_managed_policies.html#AmazonEC2ContainerRegistryReadOnly) to the IAM role we create. This will allow us to get the ECR tokens

2. we add a background process that will run the docker login command with the ECR tokens every X hours, to make sure the tokens are never expired.

for 2. I'm not sure if we need to run that on every host, or if we just need to do it on one leader, and swarm will automatically replicate it across, I'll need to do some research first.

## Questions:
1. Would this work for you?
2. Would Read Only be fine, if someone needs more then read only, they will need to manually update the policy/role, which might not be ideal.

**Comment by ambrons** [comment]
@kencochrane 

I agree with the Cloud Formation question for ECR support as well as adding the `AmazonEC2ContainerRegistryReadOnly`.  I don't think they would need read / write access unless  you had a docker running into the swarm as say part of a CI tool with Docker in Docker and you built a docker image and wanted to push it to ECR after it was created.  I'm not sure how often that would be the case.

I know for our immediate needs ReadOnly would be fine.

As for the part about ECR tokens every x hours.  I'm still not sure that's necessary.  I've include the following documentation that shows the interaction between EC2 Instance, InstanceProfile, Role, Policy, and an application running on EC2.

http://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html

I know for the applications we've written-to-date that run in Docker on top of an EC2 Instance we do know have to handle tokens or the like as long as the EC2 Instance InstanceProfile + Role includes the proper permissions in a Policy associated to the Role.  Which is why I was surprised when I added the ECR Policy manually to the generated Role from the Docker for AWS cloud formation it didn't _just work_.

My assumption is that the internal docker containers running for aws meta and console access might be interfering with the natural behavior of the EC2 Instance resource access model.

All of that said, regardless of the underpinnings I believe the net result would be the same.  The user of docker swarm would not have to any additional steps to pull from ERC and it would just work assuming the above AmazonEC2ContainerRegistryReadOnly policy was associated with the Docker for AWS Role.

**Comment by kencochrane** [comment]
> I agree with the Cloud Formation question for ECR support as well as adding the AmazonEC2ContainerRegistryReadOnly. Let that be a Parameter of the template. I don't think they would need read / write access unless you had a docker running on the swarm as say part of a CI tool with Docker in Docker and you built a docker image and wanted to push it to ECR after it was created. I'm not sure how often that would be the case.

> I know for our immediate needs ReadOnly would be fine.

OK, we can start with read only for now, and see where that gets us.

> As for the part about ECR tokens every x hours. I'm still not sure that's necessary. I've include the following documentation that shows the interaction between EC2 Instance, InstanceProfile, Role, Policy, and an application running on EC2.

> http://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html

> I know for the applications we've written-to-date that run in Docker on top of an EC2 Instance we do not have to handle tokens or the like as long as the EC2 Instance InstanceProfile + Role includes the proper permissions in a Policy associated to the Role. Which is why I was surprised when I added the ECR Policy manually to the generated Role from the Docker for AWS cloud formation it didn't just work.

The IAM role will allow you to run the aws ecr commands to get the token, but you will still need to run `aws ecr get-login` every 12 hours, or else the token will expire. It doesn't automatically update your token for you. see http://docs.aws.amazon.com/cli/latest/reference/ecr/get-login.html for more details.

We could use something like this: https://github.com/awslabs/amazon-ecr-credential-helper which is a docker credential store, which will automatically handle collecting the ECR tokens for you, without the need for running docker login, etc.

**Comment by ambrons** [comment]
@kencochrane 
> The IAM role will allow you to run the aws ecr commands to get the token, but you will still need to run aws ecr get-login every 12 hours, or else the token will expire. It doesn't automatically update your token for you. see http://docs.aws.amazon.com/cli/latest/reference/ecr/get-login.html for more details.
>We could use something like this: https://github.com/awslabs/amazon-ecr-credential-helper which is a docker credential store, which will automatically handle collecting the ECR tokens for you, without the need for running docker login, etc.

I guess amazon-ecr-credential-helper is similiar to this which runs on each EC2 Container for docker: https://github.com/aws/amazon-ecs-agent

Okay so now that my magic has been debunked.  Is the plan to have the Cloud Formation for SWARM set all of this up for ECR users?  Such that if they choose to include the ECR Role Or Docker Repository in their Cloud Formation that the created EC2 instances will keep the tokens refreshed?  

So to recap, now that I'm on the same page with you.  I don't have an issue using `aws ecr get-login` on my local system over an SSH tunnel to the docker swarm manager to initially create deployments and stacks.  That works as advertised.  What isn't solved is the issue of after 12 hours the token used during the install expiring such that the scheduler gets borked if the service is updated or rebalanced post 12 hours.

So ignoring the IAM Role for a minute I guess the real issue is keeping the authentication refreshed post service or stack create / deployment.

Is there a way to auto configure / deploy https://github.com/awslabs/amazon-ecr-credential-helper as a global container along side the other global docker containers per cloud formation instance such that this all just works?

Having never used the docker credential helper before It appears you might not need AWS KEY / SECRET if you use the `AWS_SDK_LOAD_CONFIG=true` for assuming the role of the instance.  Or perhaps that's in addition to AWS KEY and SECRET.

**Comment by kencochrane** [comment]
short answer. Yes, the goal is to allow someone to pick the "enable ECR support" and it will deal with the IAM profile, and keeping the Token up to date.

**Comment by DanielMaier-BSI** [comment]
Hi,

first of all, thanks for this cloudformation template. This is really awesome. 

We are also very interested in this feature. Did you already experiment with this or is there even any schedule when ECR integration gets implemented?

Many thanks
Daniel

**Comment by jeffhuys** [comment]
No ECR support is a really big problem for us. We want to move from ECS to Docker for AWS, but would like to have all our build scripts remain mostly the same...

**Comment by kencochrane** [comment]
@jeffhuys it is on the road map, current estimate is for it to arrive with 17.06-ce

**Comment by kencochrane** [comment]
It has been a while since I have commented on this issue, so I think it is time for an update, to let everyone know where we are with this issue. 

Before I get into the current status, I'll give a little background.

The way ECR handles auth is by using IAM roles for your servers, that allow you to get short lived Auth tokens, which can be used to docker push/pull your images. 

There are two common ways for doing this. you can do `aws ecr get-login` or use the `amazon-ecr-credential-helper`. This will get your auth tokens and put them in the right location so that docker can use them, just like any other credential.

Docker for AWS uses Swarmkit for our orchestration, and when you deploy your service, you can use the `--with-registry-auth` flag which will take the ECR credentials you have locally and pass them to the workers who are going to run your service containers, so they have access to those credentials.

This all works as it should, and if you never need the ECR auth tokens again, there would be no problem. Unfortunately, you will most likely need this token in the future. Even if you are not going to be doing a redeploy, you would need it, if a host goes down, and your containers get rescheduled to a new node. If this new node doesn't have the required images already loaded in it's image cache, it will need to pull down those images from ECR. If this happens after the ECR auth token expired, the image pulls from ECR will fail with invalid credentials. 

Another important thing to consider, is that there are two different parts to docker, the server and the client. The server is the docker daemon that runs on the host and manages the containers, etc. The client is a CLI binary which interacts with a given docker daemon, it might be locally, or it could be on a remote server. Docker Auth credentials are handled by the docker client, and the docker server doesn't have access to them. Because of this, we can't use `aws ecr get-login` or `amazon-ecr-credential-helper` on swarm workers to get new auth tokens, when the current ones expire.

One work around would be to get a new auth token and then do a `docker service update` every X hours before the token expires, but this would cause a rolling update of the service to update the registry creds in all worker nodes, which isn't ideal. We have some other ideas as well, but they are also not very good. What ever solution we finally choose, we want to make sure it is something that works well for everyone, and won't make things worse in the long run. 

The only viable solution is to make changes to the way docker swarm handles registry auth credentials. We are currently working with that team, to come up with a solution that works not just for ECR but for all registries. If you want to track the progress feel free to follow this issue: https://github.com/moby/moby/issues/24940

Until those changes are added to swarm, we are currently blocked. I'll be sure to update this issue, with our progress, to let everyone know where we stand.

**Comment by blaketastic2** [comment]
As a work around for this(I didn't want to install aws cli on the manger node), I was able to utilize the aws cli on the "guide-aws" container that is started on my manager node. So by running the following, I can capture the login command:

docker exec -it guide-aws sh -c 'aws ecr get-login --region us-east-1'

Then, I've set up a cron to run on the manager every 6 hours to re-login the manager.

Would love some thoughts on this...

**Comment by kencochrane** [comment]
@blaketastic2 did this approach work for you? 

At first glance, I don't think it will work because running the `aws ect get-login` will not update the password that is tied to the service. Swarm takes the password on service create, encrypts it and stores it in service config. If that password changes the service will need to be updated in order for the new password to be set.

**Comment by blaketastic2** [comment]
As far as I can tell, this cron works:
`$ docker exec -it guide-aws sh -c 'aws ecr get-login --region us-east-1' | tr -d '\r' > login && ./login`

Followed by this service update:
`$ docker service update --with-registry-auth serviceA`

**Comment by kencochrane** [comment]
@blaketastic2 yes, if you run the `service update` after the `aws ecr get-login` that should work. It isn't ideal since it will cause a rolling update of the service. You will also need to remember to update every service that needs the new credentials. 

If it works for you, then 👍

**Comment by friism** [comment]
From talking to @diogomonica, this is how progress will be made on this problem:

 1. Docker will get plug'able external secret stores
 2. `--with-registry-auth` transitions to using docker secrets for storing registry creds (right now I think it's just embedded in the service def)
 3. We implement an EC2 instance metadata pseudo-secret store
 4. That gets wired up so that swarm dynamically reads ECR registry creds as a secret, with that secret provided by the pseudo-secret EC2 instance metadata secrets plugin.

**Comment by RehanSaeed** [comment]
Any guestimates, rough or otherwise when we might be able to see this option in the cloud formation template?

**Comment by ambrons** [comment]
@friism @kencochrane this looks good to me, thanks for the update, much appreciated!

**Comment by destebanm** [comment]
Hi @friism, do you know when the solution you explained will be available?

@kencochrane, is there any way to update the service without doing a rolling upgrade? I mean only update credentials every x hours but without causing a rolling update of the service. This would be a good workaround while the final solution is developed.

Thanks!

**Comment by kencochrane** [comment]
As of right now, I do not know of a way to update the credentials without causing a rolling update of the service.

**Comment by destebanm** [comment]
OK, @kencochrane thanks!
So I suppose that we have to wait until solution explained by @friism is available in EC2 instances and docker swarm.
Meanwhile we have to do a rolling update after login in ECR, because is the only way to be sure that credentials are spread and new EC2 instances run the service succesfully

**Comment by joostaafjes** [comment]
@blaketastic2 I also don't want to install aws cli, but when doing 

`docker exec -it guide-aws sh -c 'aws ecr get-login --region eu-west-1'  `

I get the following error: 

`An error occurred (AccessDeniedException) when calling the GetAuthorizationToken operation: User: arn:aws:sts::xxxxxxxxxxxxx:assumed-role/Docker-xxxxx-swarm-ProxyRole-xxxxxxx/i-xxxxxxxxxxxxxxxxxxx is not authorized to perform: ecr:GetAuthorizationToken on resource: *`

Did you also do something else to get this working?

thx

**Comment by joostaafjes** [comment]
@blaketastic2 already found out, also had to add the policy AmazonEC2ContainerRegistryReadOnly to the role.

thx for the elegant solution. I slight modified the oneliner to prevent writing a file...

so my cron looks like:

`0 */6 * * * eval "$(docker exec -it guide-aws sh -c 'aws ecr get-login --region eu-west-1 --no-include-email'| tr -d '\r')" && docker service update --with-registry-auth main_xxxx`

**Comment by tomalok** [comment]
would that crontab run on moby (as root or docker?), in the shell-aws container (presumably as docker), or elsewhere?

also, are the credentials tied to the specific service, or would updating one token service which wouldn't be affected by restarts be sufficient to distribute the credentials?

very much looking forward to a more elegant and permanent solution to this problem...

**Comment by FrenchBen** [comment]
@tomalok the `guide-aws` container is already running - The cronjob specified simply runs the command in the container. Since the container is available to your current user, you can simply set it up without being root.

**Comment by tomalok** [comment]
@FrenchBen yes, I understand that the `aws ecr get-login` command is running in `guide-aws` container.  What I'm after is where `crond` is running?  It's not set up by default on the manager node's `shell-aws` container.  (FWIW, it appears that `crond` runs in the `guide-aws` container for doing cleanup, etc.)

If I patch the CloudFormation template (around where the manager and worker instances are having their /etc/docker/daemon.json set up and /home/docker permissions set) to add a crontab entry, presumably it would be the non-containerized `crond` on the moby instance itself that would be running the ECR auth cron job.

Probably wouldn't be a bad idea to have the CloudFormation template run `aws ecr get-login` (via `guide-aws`) to get the credentials in place from the get-go.

Another question is whether or not the credentials would need to be for the docker user, the root user, or both -- for the case where a new worker node comes up and needs to pull an ECR image for a global service.

**Comment by FrenchBen** [comment]
@tomalok if you're going to modify the template, why not add your own container that gets called in the cronjob - You can bind mount the proper locations for the credential token to be updated. 
For example, something like: https://hub.docker.com/r/behance/ecr-login/

**Comment by tomalok** [comment]
At the moment, the only alterations I'm doing to the CloudFront template is adding one line to the end of each of the UserData scripts for the manager and worker nodes, which pretty much just pulls an install script down from my S3 bucket and executes it.
```
wget -qO- http://s3-us-west-2.amazonaws.com/my-bucket/install-refresh-ecr-auth.sh | sh -ex
```

The install script does the rest of the heavy lifting:
- writes the `refresh-ecr-auth.sh` script
- executes it once to set initial ECR credentials
- updates the crontab to call the `refresh-ecr-auth.sh` script once every 8 hours

The `refresh-ecr-auth.sh` script merely
- gets the docker login command via
```docker exec guide-aws aws ecr get-login --region $AWS_REGION --no-include-email```
- `eval`s the command twice, once for root, and once for the docker user
```eval $ECR_LOGIN; su docker sh -c "eval $ECR_LOGIN"```

Because the moby instance's `/home/docker` is already bind mounted inside the `shell-aws` container, the appropriate credentials are already populated in ~/.docker/config.json when you SSH in to the node.

I'm still hand-adding the AmazonEC2ContainerRegistryReadOnly policy to the Worker and Proxy roles, but I'm assuming that _should_ be relatively straightforward to add to the CloudFormation template, too.

**Comment by FrenchBen** [comment]
@tomalok Sounds like a good approach - any chance you'd consider sharing the (edited) script via a gist?

**Comment by tomalok** [comment]
Adding AmazonEC2ContainerRegistryReadOnly policy to the Worker and Proxy roles was rather straightforward, works nice.

I'll probably do a little cleanup and make the template/patch, and install script available via S3 somewhere.

**Comment by tomalok** [comment]
The patch (vs. 17.09.0-ce-aws1) and the resultant patched CloudFormation template are available at [http://tomalok-d4aws.s3-website-us-west-2.amazonaws.com/](http://tomalok-d4aws.s3-website-us-west-2.amazonaws.com/).  Also included is a link to the install script source for pre-deployment review.

"Works for me."

**Comment by mRoca** [comment]
Hello, I've created an image to fix this problem, based on the same solution as https://github.com/docker/for-aws/issues/5#issuecomment-336250817 : a service will run on a manager, doing a `$(aws ecr get-login ...)` and a `docker service update --with-registry-auth ...` for each service using ECR each 4 hours. 

See https://github.com/mRoca/docker-swarm-aws-ecr-auth , in theory you just have to create the service (see the README) on your aws swarm stack.

**Comment by tomalok** [comment]
FWIW, I'm working on an updated/enhanced template patch vs. 17.12.0-ce-aws1

**Comment by FrenchBen** [comment]
@tomalok If you'd like, I can add the ECR policy into the default template - Users will only need to install the ECR refresher manually. 
`wget -qO- http://s3-us-west-2.amazonaws.com/tomalok-d4aws/install-refresh-ecr-auth.sh | sh -ex`
This will prevent you from having to keep the template up to date and will only require the installer refresher to be added.

Additionally we could add the installer script to this repo, allowing people to contribute, and wrapping it in a container. WDYT?

**Comment by tomalok** [comment]
@FrenchBen - here's my "TODO" list for the next iteration of the CloudFront patch...
- [x]  Base it on the latest version of the CloudFormation template.
- [x]  Embed scripts/files into the CloudFormation template, no longer need to pull the install script from my S3 bucket or elsewhere.
- [ ] Make the ECR Auto-Auth feature (and other features) optional when deploying/updating the CloudFormation template.
- [ ] Adding/removing Proxy and Worker nodes optionally updates Route53 A records.  Likely based on or using https://github.com/30mhz/autoscaling-route53-lambda?
- [ ] Option to disable the creation of an external ELB.  Not sure yet  how the `l4controller-aws` container will react to that not being there, but for small numbers of small nodes, DNS + Route53 health checks + Docker Swarm's mesh is often sufficient.  (One ELB costs more than (3) reserved t2.micro's...)

I've moved stuff over to https://github.com/tomalok/docker4aws, master branch is currently the old patch...  tomalok2 branch is where I've (slowly) been working on the TODO stuff.  Of course, would be nice if some/all of these tweaks ended up going official, someday...  ;)

**Comment by FrenchBen** [comment]
@tomalok Happy to include the policy as part of the deployment (doesn't seem to hurt anything)
* [ ] Base it on the latest version of the CloudFormation template.

For the rest, it seems easier to build an ECR container installer, so that you can install/enable/disable this feature without requiring a ton of work on the users' side. WDYT?

The ELB flag is an interesting one and more than likely applies to the deployments with existing VPC. The LB would need some tweaks, which can be made against its source:
https://github.com/docker/infrakit/tree/master/pkg/provider/aws/plugin/loadbalancer

**Comment by tomalok** [comment]
@FrenchBen with regards to the ECR policy, if it's not being used (ECR Auto-Auth turned off), then it's probably best to not have the policy -- "least privileges" and all that.  Would be even better if it were configurable to permit specific ECR resources, instead of just "*".

I'm not quite sure I understand what you mean by an "ECR container installer"?  The way I've been proceeding, the option to disable/enable/configure this (and other) features is in the CloudFormation template, so the user would choose/configure when deploying or updating the stack...  Installation/setup/updates are pretty much a one-stop-shop at that point, instead of having to do something more to enable/configure any additional feature(s).

**Comment by FrenchBen** [comment]
@tomalok wrt to the installer/config, it's to  tell the user to run 1 container to do the work that's done within the CloudFormation template. This allows a user to easily enable/disable the ECR script.

Having a dynamic policy, seems a bit more work. Since there is no auth, does it really become a security issue to allow the instance access?

**Comment by Mobe91** [comment]
@tomalok First of all, thanks for providing this script.

But I have a weird issue with it. I set it up on all my manager and worker nodes and apparently it works since I am able to `docker pull` images from every ECR repository when SSHing into one of my instances without performing any manual `docker login`.
The strange thing is, I just happened to update my cluster to a new worker instance size via a CloudFormation stack update. This naturally does a rolling upgrade on my worker nodes. Everything worked fine but then none of my services that use an image from one of my ECR repositories could be restarted on the new nodes.
When I run `docker service ps` on these services, it always says `No such image` which usually means "I cannot access the ECR repository". When I do a manual `docker pull` on a node that is targeted by a service replica deployment, the deployment suddenly starts working.

Any ideas about this?

**Comment by tomalok** [comment]
Sorry for the delay...  I've been sidetracked on a bunch of things...

@Mobe91 I haven't yet tested the case of updating instance size...  it *should* work, because the instance UserData set up should take care of the initial needfuls.  I'll give it a test the next time I have a spare moment or three.

Also, @FrenchBen, etc. because of all the sidetracking, and the fact that there's been a new set of Docker for AWS templates released in the interim, I'm considering a new approach...  Leaning towards reimplementing as a script that downloads a specified template (new VPC vs. existing VPC, edge vs. stable, etc.), loads the CloudFormation JSON into a native object for modification (adding/removing features), and then outputs the new resulting template.

**Comment by tomalok** [comment]
@Mobe91 -- found some time to test tonight, although my Docker for AWS swarm is running 2017.12.0-ce-aws1 with a newer, more self-contained ECR auto-authentication patch applied (no more dependency on the S3 bucket).

Bumped my swarm up to t2.small, everything went okay, `~docker/.docker/config.json` auths had been freshly updated on all the new management nodes (I have no worker nodes at present)...  But then realized I didn't have a service running out of my ECR registry.

I `docker service create ...`'d a service from my ECR with replicas on all nodes, and then resized my swarm back down to t2.micro.  The first node that came online did not immediately start the service (because there were already enough replicas in operation), but when autoscaling removed an "old" instance, a replica of the service started there just fine, automagically without intervention.

Maybe the problems you experienced has something to do with the older version and/or a not-yet-ready dependency or weird race condition... or if this happened on a worker node, it's possible there's something different on those?  That's another situation I haven't tested...

**Comment by Mobe91** [comment]
@tomalok Thanks for taking the time to test this.
I discovered that this is not an issue with instance size change but it also happens when I simply increase the number of service replicas via `docker service update` for an existing service. The scaling fails on all nodes with `no such image` that do not yet have a running instance of the same service (i.e. no local image is present). This happens on both manager and worker nodes.
I have attached the [CloudFormation Template](https://github.com/docker/for-aws/files/1922591/docker-swarm-aws-internal-cf-template.zip) that I used to create the swarm. The installation of the ECR refresher happens in lines 1588 and 2215. Can you spot any issue with that?

**Comment by tomalok** [comment]
Apart from pulling the script from a different S3 bucket (with an additional comment at the top *BEFORE* the shebang `#!` line -- probably not the best place for a comment, but since it's piped into `sh -ex` shouldn't hurt anything...), I'm not seeing anything obvious other than it's patching 18.12.1-ce-aws1 instead of 18.12.0-ce-aws1.

Testing by adding a worker node to my swarm...  Although I can't ssh into that node (at least from outside of the security group), I see from the EC2 console log that it appears to have done the `aws ecr get-login` okay.  Bumping my service's replicas up a notch...  And it's running fine...
```
~ $ docker service update --replicas 4 jn-svc
jn-svc
overall progress: 4 out of 4 tasks
1/4: running   [==================================================>]
2/4: running   [==================================================>]
3/4: running   [==================================================>]
4/4: running   [==================================================>]
verify: Service converged
~ $ docker service ps jn-svc
ID                  NAME              IMAGE                                                        NODE                                          DESIRED STATE       CURRENT STATE            ERROR               PORTS
pcru1f5seswg        jn-svc.1          000000000000.dkr.ecr.us-west-2.amazonaws.com/jn/svc:latest   ip-172-31-99-999.us-west-2.compute.internal   Running             Running 8 hours ago
isbv94cfkblg         \_ jn-svc.1      000000000000.dkr.ecr.us-west-2.amazonaws.com/jn/svc:latest   n6y13g7qfru5pbnb75hfuyed6                     Shutdown            Running 8 hours ago
wk87ry9qyscc        jn-svc.2          000000000000.dkr.ecr.us-west-2.amazonaws.com/jn/svc:latest   ip-172-31-99-99.us-west-2.compute.internal    Running             Running 8 hours ago
qk85rj74xzxn         \_ jn-svc.2      000000000000.dkr.ecr.us-west-2.amazonaws.com/jn/svc:latest   75ug0racy5kjfp545k24oou73                     Shutdown            Running 8 hours ago
5qc8wbexhvxu        jn-svc.3          000000000000.dkr.ecr.us-west-2.amazonaws.com/jn/svc:latest   ip-172-31-9-999.us-west-2.compute.internal    Running             Running 8 hours ago
tmf4xrz4rbgl         \_ jn-svc.3      000000000000.dkr.ecr.us-west-2.amazonaws.com/jn/svc:latest   rsttef2xj2ls6dbgvbzsxuzba                     Shutdown            Running 8 hours ago
p81q4844e5fu        jn-svc.4          000000000000.dkr.ecr.us-west-2.amazonaws.com/jn/svc:latest   ip-172-31-99-9.us-west-2.compute.internal     Running             Running 14 seconds ago
```

**Comment by Mobe91** [comment]
@tomalok I guess you created the service using `--with-registry-auth`? Then I think [this issue](https://github.com/moby/moby/issues/33929) explains it all...

I just tried this again. When I create a service on a 1 node swarm using `docker service create --with-registry-auth` everything works fine. Immediatly after that, I spawn an additional swarm node and scale up the service to 2 replicas. Again, everything works fine - but that is just because the credentials stored with the service at the time of its inital creation are still valid.
This all falls apart when the service credentials expire and I guess this is the problem that am experiencing all the time, e.g. during cluster update. Of course, your ECR credential refresher script cannot do anything about this because swarm only considers the stored service credentials instead of the up-to-date client credentials. 😞 

This also explains why `docker pull`ing the images worked for me while, at the same time and on the very same node, swarm reported `No such image`.

I don't see how this is supposed to work at all...

**Comment by tomalok** [comment]
Testing again, it's been a couple days since I downsized the service and nodes to 3.  Instead of adding a node and another replication, I terminate one of the node instances entirely...  So, no opportunity to re-specify `--with-registry-auth`...

Starting by terminating a node that's not the master node...  Container 3 came up on one of the remaining two nodes...  Though, of course, the image was already on that node.  Verified that the replacement node currently does not have the image, but it does have credentials in its .docker/config.json.  Doesn't seem to be a way to `docker service update --mode global` so I added another replica, and with that, I believe I have finally reproduced your problem.

I have some ideas how I might be able to work around this...

[moments later...]
It appears to have sorted itself out while I wasn't looking, after initially complaining that it couldn't find the image... But upon closer inspection, it just chose a different node to put the replica on.

@Mobe91 is your service started with `--replicas _n_` or `--mode global`?

**Comment by tomalok** [comment]
I think I've solved it, at least for services currently running in the swarm -- I've updated the script at https://s3-us-west-2.amazonaws.com/tomalok-d4aws/install-refresh-ecr-auth.sh

The trick is that after the ECR credentials are refreshed, the refresh script immediately does `docker service update --with-registry-auth` for every service in the swarm.  The only thing this update does (I think!) is provide updated per-service credentials to all swarm docker daemons -- it shouldn't trigger any image updates, etc.

@Mobe91 - give the updated install script a try, and let me know if it works for you (or if you notice any other side effects).

**Comment by Mobe91** [comment]
@tomalok I had the same idea and ran `docker service update --with-registry-auth` on every service before I upgraded my swarm to 18.03.0-ce yesterday. 😀 Sadly, it did not have any effect, the repositories could still not be accessed by the new nodes... So it seems that the stored credentials are not updated by this command.

**Comment by tomalok** [comment]
@Mobe91 I think it needs to be done right after the new node joins the swarm, not before.  The `docker service update --with-registry-auth` will only update the containers' credentials on the existing nodes, and new nodes do not automagically receive this information as they come up.

**Comment by tomalok** [comment]
@Mobe91 could you try updating to the following template?

https://s3-us-west-2.amazonaws.com/tomalok-d4aws/18.03.0-ce-aws1-tomalok1.tmpl

...while this does not incorporate most of things on my TODO from my February post, here's what it does contain...

- [x] the template is generated from a `tweak-template` script, which pulls the CloudFormation template from a specified URL, and applies the modifications.  This makes it much easier to update when a new official Docker for AWS template is released, or to generate "no VPC" and/or "edge" variants.
- [x] updates the LaunchConfiguration names so that AutoScaling is updated when patch version changes, instead of just being updated when Docker for AWS version changes.
- [x] embeds the most recent `install-refresh-ecr-auth.sh` script in the template itself -- there is no longer a dependency on S3 to pull the install script.
- [x] the `refresh-ecr-auth.sh` now does `docker service update --with-registry-auth` for every service that exists in the swarm -- once when the node's instance is instantiated, and then once every 8 hours.

**Comment by Mobe91** [comment]
> I think it needs to be done right after the new node joins the swarm, not before. The docker service update --with-registry-auth will only update the containers' credentials on the existing nodes, and new nodes do not automagically receive this information as they come up.

I thought the point of storing the credentials in the service definition is that nodes can retrieve this information from a central location (from state maintained by the manager nodes) whenever they require it.
But clearly, I don't know exactly so I will give your updated script a try 😄

**Comment by Mobe91** [comment]
@tomalok I just did a complete cluster update with your updated script but it did not make a difference. I guess this means that `docker service update --with-registry-auth` simply does not update the service credentials.

**Comment by tomalok** [comment]
I know this may sound weird, but I'd like to get my stuff to fail too...  ;)

@Mobe91 - would you be willing to share some deeper information about how your Docker for AWS is set up?  (perhaps off-thread?)  Looking for things like...

* EC2 region vs. ECR region
* number of managers, number of workers, and their instance types
* services running...
  * how many?
  * based on what image(s) (the `FROM` in your services' `Dockerfile` -- presuming they're ultimately based on some public image)
  * whether or not the services were initially started with `--with-registry-auth`
  * if the services were started with `--replicas` or `--mode global`
  * attached volumes/networks?

thanks!

**Comment by Mobe91** [comment]
@tomalok Sure, anything that helps :)

- Region: eu-central-1
- 3 Managers of type t2.micro (though I always increase this to 5 before doing an update), 2 workers of type t2.large
- Services:
  - I have 10 services running. 9 of those services use images stored in ECR repositories - all of them fail to fetch the image after an update or when they are rescheduled to a fresh node (after enough time that would render any creational auth token expired). The 1 other service is the [swar visualizer](https://hub.docker.com/r/dockersamples/visualizer/) that always starts without problems. My services are not restricted to run on worker nodes and in fact, some of them do run on manager nodes (apart from visualizer).
  - The most trivial service is based on nginx:1.13.6-alpine image. I am just adding a website layer to the image and store the resulting image in ECR.
  - All services/stacks are started via `docker stack deploy --with-registry-auth -c docker-compose.yml` - without the `--with-registry-auth` they would not be able to fetch the image.
  - All services use `replicas` deploy modes though most of them only have 1 replica at the moment but my mentioned website service has 2 replicas, for example.
  - None of my services has any explicit networks attached, and I think only 1 of them has a volume configured.

After I updated the swarm, 9 services were not able to fetch their image. They were up long enough that any creational ECR token would have been expired. I then ran `docker service update --with-registry-auth` on each of them in sequence and they all started to deploy flawlessly.

**Comment by tomalok** [comment]
@Mobe91 - thanks for the info, a couple more questions...  :)
* when you do an upgrade, it would seem that you're actually updating the cloud formation stack 3 times?  first to increase managers to 5, then to update the Docker for AWS, and finally to reduce managers to back to 3?
* was there an issue you were having with just one cloud formation update to update Docker for AWS?  the update should temporarily add a fourth manager node that becomes active as necessary to preserve 3 masters during 100% of the entire update...  i would assume that a 5-manager cluster would temporarily bump up to 6.
* would you be willing to send me your redacted `docker-compose.yml`?  (probably the easiest way to get a list of any constraints, options, modes, health checks, limitations/reservations, placement prefs, update/restart/rollback settings, etc.)

**Comment by Mobe91** [comment]
@tomalok 
- No I am just modifying the ASG manually for this purpose and set max and desired to 5. I align the cloud formation template parameters with the manual changes - so I am only doing 1 CF stack update an after that I change back the ASG manually. Other than that, I am not introducing any manual changes to my AWS infrastructure - so far I had no problems with this.
- That is right, with 3 managers the ASG is increased to 4 instances before the rolling upgrade and likewise, with 5 managers it is bumped to 6. And yes, I had an issue once were my swarm somehow lost quorum during stack update with 3 (4) managers and since then I am using caution and increase to 5 before an update. I am not sure what caused the loss of quorum back then and I also never tried again with 3 managers. But you are right, in theory it should work with 3 managers.
- I have attached the [docker-compose.yml](https://github.com/docker/for-aws/files/1942564/docker-compose.staging.merged.zip) that I use for my staging website service. Nothing fancy there as far as I can see. There is also no health check configured for the docker image derived from nginx. Please excuse the weird ordering in the YML but the compose file is generated by combining to other compose files via `docker-compose -f common.yml -f staging.yml config` (my current way of handling different stages with common aspects).

**Comment by tomalok** [comment]
@Mobe91 - I think I figured out why this is not working, at least when a new node comes up (It probably starts working ~8h later when the first cron fires, though).

When the instance's UserData is being executed, while `docker` is in a state capable of getting the ECR creds, the node has not yet joined the swarm.  So, when the script asks or the list of services that need to be updated `--with-registry-auth`, it is unable to do so.  From the instance's system log...
```
[...]
Login Succeeded
+ docker service ls --format {{.Name}}
Error response from daemon: This node is not a swarm manager. Use "docker swarm init" or "docker swarm join" to connect this node to swarm and try again.
 [ok]
 * Starting chronyd ... [ok]
[...]
```
It should be possible to have the `docker service` portion of the script wait in the background until the node joins the swarm.

I'm also thinking that it may be sufficient to only have the ECR auth stuff on manager nodes.  Will test that theory after solving the above problem.

**Update** - Rolled back the stack template to stock 18.03.0-ce-aws1.  Services all came up okay.  This is, because, I believe the cron job which was running on the previous nodes had seeded the ECR auth and kept those services updated every 8 hours.  ECR credentials expire in 12 hours, so I'll need to wait at least that long to try moving from the stock template back to the patched template, whereupon I expect the running services not be started.  I'll go back down to the stock template, restart the services, and wait another 12 hours or so before attempting an updated patched template that waits in the background for the node to join the swarm and then updates `--with-registry-auth`.

**Comment by tomalok** [comment]
@Mobe91 - here's an updated template, which waits (up to 60s) for the node to join the swarm before updating all services `--with-registry-auth`.  It seems to take about 10-11s for a node to join the swarm, so this should be plenty of time.

https://s3-us-west-2.amazonaws.com/tomalok-d4aws/18.03.0-ce-aws1-tomalok2-rc1.tmpl

...please put it through the paces and let me know how it goes.

Additionally, this version logs ECR auth and service update progress to the instance's system log, and eliminates some things -- `refresh-ecr-auth.sh` isn't needed on Workers, and we only update ECR auth for the docker user (root user never used it).

I've also realized there's probably a couple more items that should be on the TODO list:
- ECR Policy tweaks
  - [x] restrict to pull-only (maybe push/pull selectable via configurable option).
  - [x] specify a list of ECR ARNs
- `refresh-ecr-auth.sh` (and configuration) tweaks
  - [x] multi-region ECR auth support (i.e. swarm's in eu-central-1, but need to access a registry in eu-west-1)
  - [x] cross-account ECR auth support (via `--registry-ids <acct1> [<acct2>...]`

**Comment by tomalok** [comment]
@FrenchBen - I have no idea how I managed to "un-assign" you...  :/

@Mobe91, et. al. - I have a new release candidate that includes the features I mentioned 8-ish days ago...

https://s3-us-west-2.amazonaws.com/tomalok-d4aws/18.03.0-ce-aws1-tomalok2-rc2.tmpl

There's a lot of significant changes between this and rc-1...

* New stack template parameters!
  * Separate, configurable Manager/Worker ECR policies...
    * selectable `pull-only`, `push-pull`, and `unspecified` ECR policy types.
    * you can specify a list of your ECR ARNs to which the ECR policy applies.
  * Configurable ECR auth login targets...
    * enables the caching and refreshing of ECR credentials for different region(s) and/or different AWS account(s) -- provided the account has granted your account access, of course.
    * an empty value disables ECR credential caching/refreshing altogether.
    * the default value of `=` expands to the current region and current account-id.
    * a list of regions (i.e. `us-west-2,eu-central-1,ap-northeast-1`) caches/refreshes ECR creds for your account-id in those regions.
    * to cache/refresh creds for other accounts' ECR registries, use a `=` (optionally after a specified region) followed by a `:`-separated list of account-ids;  `=123456789012` caches/refreshes ECR creds for the 123456789012 AWS account in the current region.  `sa-east-1=:210987654321` caches/refreshes for your account and the other specified account in sa-east-1.
    * the contrived example target list `=,us-east-2=123456789012:112233445566,us-west-2=:210987654321` would cache/refresh ECR creds for your account-id in the current region and us-west-2, for two other account-ids in us-east-2, and one other account-id in us-west-2.
* New stuff in `install-refresh-ecr-auth.sh`...
  * echoes progress (viewable in EC2 instance system logs).
  * sanitizes/expands ECR auth login targets.
  * only installs `refresh-ecr-auth.sh` if ECR auth login target(s) are specified.
* New stuff in `refresh-ecr-auth.sh`...
  * logs to STDERR (for EC2 instance system logs) **and** syslog (/var/log/syslog*).
  * refreshes ECR credentials for all specified ECR auth login targets.

I think the only things between rc2 and an actual release are a proper `README`, a pull-request/merge, and the unveiling of its GitHub repository!

**Comment by Mobe91** [comment]
@tomalok Thank you for the update. Just wanted to let you know that I have not forgotten about this and I will try the new script as soon as possible - actually I wanted to do this now but I discovered that I cannot update my cluster at the moment because I require some other third-party fixes before. So I am afraid this will have to wait a bit longer.

**Comment by Mobe91** [comment]
@tomalok Tried the new template!

I have good news and bad news ;)

**Bad news** first: My cluster somehow got destroyed.
After the CF update, the leader node as reported by the DynamoDB entry reported: The swarm does not have a leader
So apparently, the swarm somehow lost consensus although the CF update did not report any errors during the rolling update procedure. Note that I upscaled to 5 manager nodes before starting the update (which increased the ASG size to 6).

I also checked the state of other manager nodes:
Manager nodes from the same AZ as the DynamoDB leader also reported "The swarm does not have a leader".
Manager nodes from different AZs reported "Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded".
However, I verified that the manager nodes from other AZs were able to ping the DynamoDB leader so network connectivity should not have been an issue.

I then tried to recover the cluster:
On the DynamoDB leader I ran
```
docker swarm init --force-new-cluster # successful, cluster state could be restored
```
I then tried to join the remaining manager nodes manually using the command returned by `docker swarm join-token manager`. I tried with 2 remaining manager nodes:
First node:
* First, run `docker swarm leave --force` - ok
* Run join command: timeout

Second node:
* `docker swarm leave --force` - timeout

I checked back with the DynamoDB leader node and found it largely unresponsive and the EC2 monitoring showed that the CPU utilization was up like crazy.

So I thought let's try to re-establish the cluster from the 2 remaining nodes that I had not run ``docker swarm leave --force` on at the time. Same siutation on both nodes:
* `docker swarm init --force-new-cluster`
* `docker swarm join-token manager` to get the join command. This returned NO OUTPUT whatsoever meaning no error and no join command. By then I was like _what the hell_... 
* I also tried `docker swarm join-token worker` --> no output

I then checked back with the DynamoDB leader and found it to be back in a responsive state but it reported "The swarm does not have a leader" again.
So I ran `docker swarm init --force-new-cluster` again on this node. This worked fine but now my cluster state was gone. Joining the remaining manager and worker nodes to the swarm worked smoothly now.

**Good news**: I ran `docker ps` on some of the manager nodes after the CF update and some had containers running for images that required access to the ECR repositories. So it _appears_ that the ECR access has worked now.

**Comment by fernandobandeira** [comment]
@tomalok It's working out for me, didn't have any problems migrating to your template. I'll keep testing and will post an update if anything happens.

**Comment by tomalok** [comment]
Thanks for the feedback, @Mobe91 ...  could you clarify whether you changed your cluster size via a CloudFormation stack update, or if you did it by modifying the autoscaling group directly (by editing the details tab)?

It's been my experience that when you add instances via the CF stack, the new ones are added one at a time; but when you edit the autoscaling group details, the new instances are spun up nearly simultaneously -- perhaps there's a race condition when multiple new nodes join at the same time?  Also, the CF stack, as a whole, is unaware that there are now additional manager nodes, and still reports the original number of nodes.

Not certain whether either is the root of the problem you're experiencing.

But, at least it's good to hear that the ECR access worked!  :)

**Comment by Mobe91** [comment]
@tomalok I changed the number of managers by modifying the autoscaling group. So you could be right but it would still be weird if that was the reason. I thought that the CF stack would be able to just consult the ASG - that it knows of - for the number of instances.

Also, if that is the problem, how should we implement actual node auto-scaling for the cluster then, e.g. a service deployment reports insufficient resources which triggers the creation of a new node?

**Comment by tomalok** [comment]
@Mobe91 Next time you want to change the number of managers, I suggest updating your existing Cloud Formation stack, and increasing/decreasing the number of managers in the parameters...  I suspect you'll have better luck maintaining a sane state for your swarm.

From what I can tell, apart from maintaining a set amount of managers/workers, the CloudFormation stack as it is doesn't currently support auto-scaling based on load or some other metric.  I suspect, however, this might be a good candidate for an additional template feature...  :)

**Comment by fernandobandeira** [comment]
@tomalok Do you know how can we add ssl certificates to the load balancer? I mean I've read the tutorial on docker for aws, and added this to my compose file:

```yml
labels:
      com.docker.aws.lb.arn: "arn:aws:acm:us-east-1:393305476528:certificate/4c2baa9e-40ec-4a2c-ba2c-53224888466e@443"
```

But the 443 port isn't using this acm certificate, I've changed it manually for now and it works fine, (what I need is SSL on load balancer port 443 targeting TCP on instance port 443), I know it's not related to this issue but I'm adding it here since I'm using your template (didn't test this on the official template), so maybe it can be an issue with your template or with support for labels in the docker compose file, or just me doing some misconfiguration, the rest is working flawlesly, tested with a lot of machines and services, replicas and things are looking good.

**Comment by tomalok** [comment]
@fernandobandeira I haven't played around with ACM and/or LBs yet -- (I'd like to be able to turn off the LB altogether, to save $ on my small swarm) -- but perhaps the swarm doesn't have the proper IAM permissions to attach the certificate?  Another thing on the to-do list...

**Comment by Mobe91** [comment]
@fernandobandeira Port 443 should be the default according to the documentation so you could try if `arn:aws:acm:us-east-1:393305476528:certificate/4c2baa9e-40ec-4a2c-ba2c-53224888466e` works. Other than that, I don't see an issue with your label - I am using it the same way in my docker-compose file and it works. You might want to verify that your service has indeed the specified label attached by `docker service inspect`ing it. Howevever, this should not have anything to do with @tomalok 's template.

@tomalok I know that auto-scaling is not supported right now but I was thinking of building it myself. So I would have to do this via cloud formation stack updates. hm.. It is currently not a priority for me but it would definitely be nice feature 👍

**Comment by Mobe91** [comment]
@tomalok Since my last comment I have attempted 2 updates of my cluster. The first went very well, i.e. the cluster did not break and all services were able to pull the images and start up fine.

The second updated showed similar results to what I described in one of my [previous comments ](https://github.com/docker/for-aws/issues/5#issuecomment-390400802). This time I did not apply any non-CloudFormation changes to the environment. 5 new nodes were started in order. After the final node started, the situation was as follows:

Node 1 and Node 2 returned error `Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded` on any swarm related command.
Using `docker swarm init --force-new-cluster` I was able to restore the cluster state on these two nodes but then `docker swarm join-token manager` returned with empty result.

Node 3 and Node 4 returned error `Error response from daemon: This node is not a swarm manager. Use "docker swarm init" or "docker swarm join" to connect this node to swarm and try again.` on any swarm related command.
Restoring the cluster state using `docker swarm init --force-new-cluster` was not possible for these nodes but `docker swarm join-token manager` returned a valid output.

Node 5 ended up as the leader of a functioning swarm with all worker nodes attached but with empty cluster state. The DynamoDB entry pointed to node 5 after the update.

I have collected and attached various [logs](https://github.com/docker/for-aws/files/2136933/swarm-logs.zip) from all nodes.
One more fact: I have a second swarm that only consists of managers and that only runs a simple nginx service and updates of this swarm always work without problems. Apart from minor differences, this swarm uses the same CF template as the other swarm.

docker version:
```
Client:
 Version:       18.03.0-ce
 API version:   1.37
 Go version:    go1.9.4
 Git commit:    0520e24
 Built: Wed Mar 21 23:05:35 2018
 OS/Arch:       linux/amd64
 Experimental:  false
 Orchestrator:  swarm

Server:
 Engine:
  Version:      18.03.0-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   0520e24
  Built:        Wed Mar 21 23:14:32 2018
  OS/Arch:      linux/amd64
  Experimental: true
```

**Comment by tomalok** [comment]
Had some time this morning, so I added the final feature for the "tomalok2" tweak version -- ACM policy management, of particular interest to @fernandobandeira, perhaps...

I've also moved my git repo over to GitLab (https://gitlab.com/tomalok/docker4aws), which includes the `tweak-template` script I've been using to patch various versions of the Docker for AWS template.  Let's call this the official "tomalok2" release.

At the time I'm writing this, it appears that the current stable release is 18.03.0-ce-aws1, and the current edge release is 18.04.0-ce-aws1.  `tweak-template` is known to work with these and earlier versions (I think 17.09.0-ce-aws1 is the earliest I've tried); we'll revisit if/when another Docker for AWS template is released -- I would have expected 18.06.0 by now...  :/

**Comment by GreatEarl** [comment]
Hi @tomalok . It looks like your solution is can be used between EC2 and ECR (I'm a bit new to all of that) We've a few EC2 instances and ECR with docker images. We use [watchtower](https://github.com/v2tec/watchtower) to update automatically but we need to re-authenticate every 12 hours. The goal is to configure continuous pulls without need to re-login. Is that achievable with your tweaks?

Best.

**Comment by tomalok** [comment]
I'm not familiar with watchtower, but yes, this patch to the Docker for AWS CloudFormation template adds ECR credential renewal.  Normally AWS ECR credentials expire after 12h, this refreshes credentials for services deployed to the Docker Swarm once every 8 hours.

**Comment by GreatEarl** [comment]
@tomalok thanks mate but I'm not sure do we use Docker for AWS (I mean EE). I thought to make a cron script for login from EC2 vm to ECR like scheduled for every 8 or 10 hours. But if watchtower regularly checks docker-repo for changes may be there is a possibility to check or to login? I'll try to ask this on their branch.
